File: lib/level/tilesets/gray_wall_tile_set.dart
import 'dart:ui';

import 'package:bloo_dot_evolutions/algo/blob_tile_painter_base.dart';
import 'package:bloo_dot_evolutions/algo/coords2.dart';
import 'package:bloo_dot_evolutions/algo/monomino_lookup.dart';
import 'package:bloo_dot_evolutions/arena/level_base.dart';
import 'package:flutter/material.dart';

class GreyWallTileSet extends BlobTilePainter {
  static const marbleFloor = 1;

  /* https://www.qb64.org/wiki/COLOR */
  final Paint basePaint = Paint()
    ..color = const Color.fromARGB(0xff, 0x54, 0x54, 0x54)
    ..style = PaintingStyle.fill;

  final Paint decoPaint = Paint()
    ..color = const Color.fromARGB(0xff, 0xa8, 0xa8, 0xa8)
    ..style = PaintingStyle.stroke
    ..strokeWidth = 2.43;

  static const cornerNWx = LevelBase.tileSize / 7;
  static const cornerNWy = cornerNWx;
  static const cornerNEx = LevelBase.tileSize - LevelBase.tileSize / 7;
  static const cornerNEy = cornerNWx;
  static const cornerSWx = cornerNWx;
  static const cornerSWy = cornerNEx;
  static const cornerSEx = cornerNEx;
  static const cornerSEy = cornerSWy;

  @override
  String get spriteSheetName => "tile_set_001.png";

  @override
  paintBlobTile(Canvas paintTo, int primeIndex) {
    paintTo.drawRect(const Rect.fromLTWH(0, 0, LevelBase.tileSize, LevelBase.tileSize), basePaint);
    switch (primeIndex) {
      case MonominoLookup.primeIndexShy:
        paintTo.drawRect(const Rect.fromLTRB(cornerNWx, cornerNWy, cornerSEx, cornerSEy), decoPaint);
        break;

      case MonominoLookup.primeIndexU:
        paintTo.drawRect(const Rect.fromLTRB(cornerNWx, -cornerNWy, cornerSEx, cornerSEy), decoPaint);
        break;

      case MonominoLookup.primeIndexKnee:
        paintTo.drawLine(const Offset(cornerNWx, -cornerNWy), const Offset(cornerSWx, cornerSWy), decoPaint);
        paintTo.drawLine(const Offset(cornerSWx, cornerSWy), const Offset(cornerSEx * 2, cornerSEy), decoPaint);
        paintTo.drawLine(const Offset(cornerNEx, -cornerNEy), const Offset(cornerNEx, cornerNEy), decoPaint);
        paintTo.drawLine(const Offset(cornerNEx, cornerNEy), const Offset(cornerNEx * 2, cornerNEy), decoPaint);
        break;

      case MonominoLookup.primeIndexL:
        paintTo.drawLine(const Offset(cornerNWx, -cornerNWy), const Offset(cornerSWx, cornerSWy), decoPaint);
        paintTo.drawLine(const Offset(cornerSWx, cornerSWy), const Offset(cornerSEx * 2, cornerSEy), decoPaint);
        break;

      case MonominoLookup.primeIndexPipe:
        paintTo.drawLine(const Offset(cornerNWx, -cornerNWy), const Offset(cornerSWx, cornerSWy * 2), decoPaint);
        paintTo.drawLine(const Offset(cornerNEx, -cornerNEy), const Offset(cornerSEx, cornerSEy * 2), decoPaint);
        break;

      case MonominoLookup.primeIndexT:
        paintTo.drawLine(const Offset(cornerNWx, -cornerNWy), const Offset(cornerSWx, cornerSWy * 2), decoPaint);
        paintTo.drawLine(const Offset(cornerNEx, -cornerNEy), const Offset(cornerNEx, cornerNEy), decoPaint);
        paintTo.drawLine(const Offset(cornerNEx, cornerNEy), const Offset(cornerNEx * 2, cornerNEy), decoPaint);
        paintTo.drawLine(const Offset(cornerSEx, cornerSEy * 2), const Offset(cornerSEx, cornerSEy), decoPaint);
        paintTo.drawLine(const Offset(cornerSEx, cornerSEy), const Offset(cornerSEx * 2, cornerSEy), decoPaint);
        break;

      case MonominoLookup.primeIndexWSE:
        paintTo.drawLine(const Offset(cornerNWx, -cornerNWy), const Offset(cornerSWx, cornerSWy * 2), decoPaint);
        paintTo.drawLine(const Offset(cornerSEx, cornerSEy * 2), const Offset(cornerSEx, cornerSEy), decoPaint);
        paintTo.drawLine(const Offset(cornerSEx, cornerSEy), const Offset(cornerSEx * 2, cornerSEy), decoPaint);
        break;

      case MonominoLookup.primeIndexWNE:
        paintTo.drawLine(const Offset(cornerNWx, -cornerNWy), const Offset(cornerSWx, cornerSWy * 2), decoPaint);
        paintTo.drawLine(const Offset(cornerNEx, -cornerNEy), const Offset(cornerNEx, cornerNEy), decoPaint);
        paintTo.drawLine(const Offset(cornerNEx, cornerNEy), const Offset(cornerNEx * 2, cornerNEy), decoPaint);
        break;

      case MonominoLookup.primeIndexW:
        paintTo.drawLine(const Offset(cornerNWx, -cornerNWy), const Offset(cornerSWx, cornerSWy * 2), decoPaint);
        break;

      case MonominoLookup.primeIndexCross:
        paintTo.drawLine(const Offset(-cornerNWx, cornerNWy), const Offset(cornerNWx, cornerNWy), decoPaint);
        paintTo.drawLine(const Offset(cornerNWx, cornerNWy), const Offset(cornerNWx, -cornerNWy), decoPaint);
        paintTo.drawLine(const Offset(-cornerSWx, cornerSWy), const Offset(cornerSWx, cornerSWy), decoPaint);
        paintTo.drawLine(const Offset(cornerSWx, cornerSWy), const Offset(cornerSWx, cornerSWy * 2), decoPaint);
        paintTo.drawLine(const Offset(cornerNEx, -cornerNEy), const Offset(cornerNEx, cornerNEy), decoPaint);
        paintTo.drawLine(const Offset(cornerNEx, cornerNEy), const Offset(cornerNEx * 2, cornerNEy), decoPaint);
        paintTo.drawLine(const Offset(cornerSEx, cornerSEy * 2), const Offset(cornerSEx, cornerSEy), decoPaint);
        paintTo.drawLine(const Offset(cornerSEx, cornerSEy), const Offset(cornerSEx * 2, cornerSEy), decoPaint);
        break;

      case MonominoLookup.primeIndexInvThreeNE:
        paintTo.drawLine(const Offset(-cornerNWx, cornerNWy), const Offset(cornerNWx, cornerNWy), decoPaint);
        paintTo.drawLine(const Offset(cornerNWx, cornerNWy), const Offset(cornerNWx, -cornerNWy), decoPaint);
        paintTo.drawLine(const Offset(-cornerSWx, cornerSWy), const Offset(cornerSWx, cornerSWy), decoPaint);
        paintTo.drawLine(const Offset(cornerSWx, cornerSWy), const Offset(cornerSWx, cornerSWy * 2), decoPaint);
        paintTo.drawLine(const Offset(cornerSEx, cornerSEy * 2), const Offset(cornerSEx, cornerSEy), decoPaint);
        paintTo.drawLine(const Offset(cornerSEx, cornerSEy), const Offset(cornerSEx * 2, cornerSEy), decoPaint);
        break;

      case MonominoLookup.primeIndexTwoNWSW:
        paintTo.drawLine(const Offset(-cornerNWx, cornerNWy), const Offset(cornerNWx, cornerNWy), decoPaint);
        paintTo.drawLine(const Offset(cornerNWx, cornerNWy), const Offset(cornerNWx, -cornerNWy), decoPaint);
        paintTo.drawLine(const Offset(-cornerSWx, cornerSWy), const Offset(cornerSWx, cornerSWy), decoPaint);
        paintTo.drawLine(const Offset(cornerSWx, cornerSWy), const Offset(cornerSWx, cornerSWy * 2), decoPaint);
        break;

      case MonominoLookup.primeIndexTwoNWSE:
        paintTo.drawLine(const Offset(-cornerNWx, cornerNWy), const Offset(cornerNWx, cornerNWy), decoPaint);
        paintTo.drawLine(const Offset(cornerNWx, cornerNWy), const Offset(cornerNWx, -cornerNWy), decoPaint);
        paintTo.drawLine(const Offset(cornerSEx, cornerSEy * 2), const Offset(cornerSEx, cornerSEy), decoPaint);
        paintTo.drawLine(const Offset(cornerSEx, cornerSEy), const Offset(cornerSEx * 2, cornerSEy), decoPaint);
        break;

      case MonominoLookup.primeIndexThreeNW:
        paintTo.drawLine(const Offset(-cornerNWx, cornerNWy), const Offset(cornerNWx, cornerNWy), decoPaint);
        paintTo.drawLine(const Offset(cornerNWx, cornerNWy), const Offset(cornerNWx, -cornerNWy), decoPaint);
        break;

      case MonominoLookup.primeIndexImmersed:
        /* nothing to do - this is identical to the blank */
        break;

      default:
        throw UnsupportedError("${(GreyWallTileSet)} cannot draw $primeIndex");
    }
  }

  @override
  IntCoords2 tileNumberToBaseSheetPosition(int tileNumber) {
    switch (tileNumber) {
      case marbleFloor:
        return const IntCoords2(0, 0);

      default:
        throw RangeError(tileNumber);
    }
  }

  @override
  IntSize2 tilePartitionSizeOnSheet(int tileNumber) {
    switch (tileNumber) {
      case marbleFloor:
        return const IntSize2(3, 3);

      default:
        return super.tilePartitionSizeOnSheet(tileNumber);
    }
  }
}

File: app/src/main/java/oy/sarjakuvat/flamingin/bde/level/tilesets/GrayWallTileset.kt
package oy.sarjakuvat.flamingin.bde.level.tilesets

import android.graphics.*
import android.util.Size
import oy.sarjakuvat.flamingin.bde.R
import oy.sarjakuvat.flamingin.bde.algo.MonominoLookup
import oy.sarjakuvat.flamingin.bde.rendition.offscreen.BlobPainterBase
import java.lang.IllegalArgumentException

class GrayWallTileset : BlobPainterBase() {
    private val basePaint = Paint()
    private val decoPaint = Paint()

    override val spriteSheetName get() = R.drawable.tile_set_001

    override fun paintBlobTile(paintTo: Canvas, primeIndex: Int) {
        paintTo.drawRect(Rect(0, 0, tileSize, tileSize), basePaint)
        when (primeIndex) {
            MonominoLookup.primeIndexShy -> {
                paintTo.drawRect(RectF(cornerNWx, cornerNWy, cornerSEx, cornerSEy), decoPaint)
            }
            MonominoLookup.primeIndexU -> {
                paintTo.drawRect(RectF(cornerNWx, -cornerNWy, cornerSEx, cornerSEy), decoPaint)
            }
            MonominoLookup.primeIndexKnee -> {
                paintTo.drawLine(cornerNWx, -cornerNWy, cornerSWx, cornerSWy, decoPaint)
                paintTo.drawLine(cornerSWx, cornerSWy, cornerSEx * 2f, cornerSEy, decoPaint)
                paintTo.drawLine(cornerNEx, -cornerNEy, cornerNEx, cornerNEy, decoPaint)
                paintTo.drawLine(cornerNEx, cornerNEy, cornerNEx * 2f, cornerNEy, decoPaint)
            }
            MonominoLookup.primeIndexL -> {
                paintTo.drawLine(cornerNWx, -cornerNWy, cornerSWx, cornerSWy, decoPaint)
                paintTo.drawLine(cornerSWx, cornerSWy, cornerSEx * 2f, cornerSEy, decoPaint)
            }
            MonominoLookup.primeIndexPipe -> {
                paintTo.drawLine(cornerNWx, -cornerNWy, cornerSWx, cornerSWy * 2f, decoPaint)
                paintTo.drawLine(cornerNEx, -cornerNEy, cornerSEx, cornerSEy * 2f, decoPaint)
            }
            MonominoLookup.primeIndexT -> {
                paintTo.drawLine(cornerNWx, -cornerNWy, cornerSWx, cornerSWy * 2f, decoPaint)
                paintTo.drawLine(cornerNEx, -cornerNEy, cornerNEx, cornerNEy, decoPaint)
                paintTo.drawLine(cornerNEx, cornerNEy, cornerNEx * 2, cornerNEy, decoPaint)
                paintTo.drawLine(cornerSEx,cornerSEy * 2f, cornerSEx, cornerSEy, decoPaint)
                paintTo.drawLine(cornerSEx, cornerSEy, cornerSEx * 2f, cornerSEy, decoPaint)
            }
            MonominoLookup.primeIndexWSE -> {
                paintTo.drawLine(cornerNWx, -cornerNWy, cornerSWx, cornerSWy * 2f, decoPaint)
                paintTo.drawLine(cornerSEx, cornerSEy * 2f, cornerSEx, cornerSEy, decoPaint)
                paintTo.drawLine(cornerSEx, cornerSEy, cornerSEx * 2f, cornerSEy, decoPaint)
            }
            MonominoLookup.primeIndexWNE -> {
                paintTo.drawLine(cornerNWx, -cornerNWy, cornerSWx, cornerSWy * 2f, decoPaint)
                paintTo.drawLine(cornerNEx, -cornerNEy, cornerNEx, cornerNEy, decoPaint)
                paintTo.drawLine(cornerNEx, cornerNEy, cornerNEx * 2f, cornerNEy, decoPaint)
            }
            MonominoLookup.primeIndexW -> {
                paintTo.drawLine(cornerNWx, -cornerNWy, cornerSWx, cornerSWy * 2f, decoPaint)
            }
            MonominoLookup.primeIndexCross -> {
                paintTo.drawLine(-cornerNWx, cornerNWy, cornerNWx, cornerNWy, decoPaint)
                paintTo.drawLine(cornerNWx, cornerNWy, cornerNWx, -cornerNWy, decoPaint)
                paintTo.drawLine(-cornerSWx, cornerSWy, cornerSWx, cornerSWy, decoPaint)
                paintTo.drawLine(cornerSWx, cornerSWy, cornerSWx, cornerSWy * 2f, decoPaint)
                paintTo.drawLine(cornerNEx, -cornerNEy, cornerNEx, cornerNEy, decoPaint)
                paintTo.drawLine(cornerNEx, cornerNEy, cornerNEx * 2f, cornerNEy, decoPaint)
                paintTo.drawLine(cornerSEx, cornerSEy * 2f, cornerSEx, cornerSEy, decoPaint)
                paintTo.drawLine(cornerSEx, cornerSEy, cornerSEx * 2f, cornerSEy, decoPaint)
            }
            MonominoLookup.primeIndexInvThreeNE -> {
                paintTo.drawLine(-cornerNWx, cornerNWy, cornerNWx, cornerNWy, decoPaint)
                paintTo.drawLine(cornerNWx, cornerNWy, cornerNWx, -cornerNWy, decoPaint)
                paintTo.drawLine(-cornerSWx, cornerSWy, cornerSWx, cornerSWy, decoPaint)
                paintTo.drawLine(cornerSWx, cornerSWy, cornerSWx, cornerSWy * 2f, decoPaint)
                paintTo.drawLine(cornerSEx,cornerSEy * 2f, cornerSEx, cornerSEy, decoPaint)
                paintTo.drawLine(cornerSEx, cornerSEy, cornerSEx * 2f, cornerSEy, decoPaint)
            }
            MonominoLookup.primeIndexTwoNWSW -> {
                paintTo.drawLine(-cornerNWx, cornerNWy, cornerNWx, cornerNWy, decoPaint)
                paintTo.drawLine(cornerNWx, cornerNWy, cornerNWx, -cornerNWy, decoPaint)
                paintTo.drawLine(-cornerSWx, cornerSWy, cornerSWx, cornerSWy, decoPaint)
                paintTo.drawLine(cornerSWx, cornerSWy, cornerSWx, cornerSWy * 2f, decoPaint)
            }
            MonominoLookup.primeIndexTwoNWSE -> {
                paintTo.drawLine(-cornerNWx, cornerNWy, cornerNWx, cornerNWy, decoPaint)
                paintTo.drawLine(cornerNWx, cornerNWy, cornerNWx, -cornerNWy, decoPaint)
                paintTo.drawLine(cornerSEx,cornerSEy * 2f, cornerSEx, cornerSEy, decoPaint)
                paintTo.drawLine(cornerSEx, cornerSEy, cornerSEx * 2f, cornerSEy, decoPaint)
            }
            MonominoLookup.primeIndexThreeNW -> {
                paintTo.drawLine(-cornerNWx, cornerNWy, cornerNWx, cornerNWy, decoPaint)
                paintTo.drawLine(cornerNWx, cornerNWy, cornerNWx, -cornerNWy, decoPaint)
            }
            MonominoLookup.primeIndexImmersed -> {
                /* nothing to do - this is identical to the blank */
            }
            else -> throw IllegalArgumentException("${GrayWallTileset::class.simpleName} cannot draw $primeIndex")
        }
    }

    init {
        /* https://www.qb64.org/wiki/COLOR */
        basePaint.color = Color.argb(0xff, 0x54, 0x54, 0x54)
        basePaint.style = Paint.Style.FILL
        decoPaint.color = Color.argb(0xff, 0xa8, 0xa8, 0xa8)
        decoPaint.style = Paint.Style.STROKE
        decoPaint.strokeWidth = 2.43f
    }

    override fun tileNumberToBaseSheetPosition(tileNumber: Int): Point {
        return when(tileNumber) {
            TileCatalog.FloorTiles.marbleFloor -> return Point(0, 0)
            else -> throw IllegalArgumentException("The number $tileNumber is not a valid value for  tileNumber in ${GrayWallTileset::class.simpleName}")
        }
    }

    override fun tilePartitionSizeOnSheet(tileNumber: Int) : Size {
        return when(tileNumber) {
            TileCatalog.FloorTiles.marbleFloor -> Size(3, 3)
            else -> super.tilePartitionSizeOnSheet(tileNumber)
        }
    }

    companion object {
        const val cornerNWx = tileSize / 7f
        const val cornerNWy = cornerNWx
        const val cornerNEx = tileSize - tileSize / 7f
        const val cornerNEy = cornerNWx
        const val cornerSWx = cornerNWx
        const val cornerSWy = cornerNEx
        const val cornerSEx = cornerNEx
        const val cornerSEy = cornerSWy
    }
}
